File: src/main/java/com/heybys/oddments/base/domain/AggregateRoot.java
package com.heybys.oddments.base.domain;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import org.springframework.data.annotation.Transient;
import org.springframework.data.domain.AfterDomainEventPublication;
import org.springframework.data.domain.DomainEvents;
import org.springframework.util.Assert;

public abstract class AggregateRoot<D extends DomainEntity<D, I>, I> extends DomainEntity<D, I> {

    @Transient
    private final List<Object> domainEvents = new ArrayList<>();

    protected void registerEvent(D event) {
        Assert.notNull(event, "Domain event must not be null");
        this.domainEvents.add(event);
    }

    @AfterDomainEventPublication
    protected void clearDomainEvents() {
        this.domainEvents.clear();
    }

    @DomainEvents
    protected Collection<Object> domainEvents() {
        return Collections.unmodifiableList(this.domainEvents);
    }

    @Override
    public boolean equals(Object other) {
        return super.equals(other);
    }

    @Override
    public int hashCode() {
        return super.hashCode();
    }
}

File: src/main/kotlin/com/heybys/oddments/base/domain/AggregateRoot.kt
package com.heybys.oddments.base.domain

import org.springframework.data.annotation.Transient
import org.springframework.data.domain.AfterDomainEventPublication
import org.springframework.data.domain.DomainEvents
import org.springframework.util.Assert
import java.util.Collections

abstract class AggregateRoot<D : DomainEntity<D, I>, I> : DomainEntity<D, I>() {

    @Transient private val domainEvents: MutableList<Any> = mutableListOf()

    protected fun registerEvent(event: D) {
        Assert.notNull(event, "Domain event must not be null")
        domainEvents.add(event)
    }

    @AfterDomainEventPublication
    protected fun clearDomainEvents() {
        domainEvents.clear()
    }

    @DomainEvents
    protected fun domainEvents(): Collection<Any> {
        return Collections.unmodifiableList(domainEvents)
    }

    override fun equals(other: Any?): Boolean {
        return super.equals(other)
    }

    override fun hashCode(): Int {
        return super.hashCode()
    }
}
